pkg/commands/image/list_filter.go

```diff
@@ -0,0 +1,122 @@
+	corev1 "k8s.io/api/core/v1"
+	"regexp"
+	"strings"
+
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+)
+
+type filter struct {
+	filterFunc func(image v1alpha1.Image, values []string) bool
+	values     []string
+}
+
+func filterImageList(images *v1alpha1.ImageList, flags []string) (*v1alpha1.ImageList, error) {
+	filters, err := parseFilters(flags)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(filters) == 0 {
+		return images, nil
+	}
+
+	var filteredItems []v1alpha1.Image
+	for _, item := range images.Items {
+		if matchesAll(item, filters) {
+			filteredItems = append(filteredItems, item)
+		}
+	}
+
+	images.Items = filteredItems
+	return images, nil
+}
+
+func parseFilters(flags []string) ([]filter, error) {
+	var (
+		filters             []filter
+		builderRegex        = regexp.MustCompile(`^builder=(.+)$`)
+		clusterBuilderRegex = regexp.MustCompile(`^clusterbuilder=(.+)$`)
+		latestReasonRegex   = regexp.MustCompile(`^latest-reason=(.+)$`)
+		statusRegex         = regexp.MustCompile(`^ready=(.+)$`)
+	)
+
+	for _, flag := range flags {
+		m := builderRegex.FindStringSubmatch(flag)
+		if len(m) == 2 {
+			filters = append(filters, filter{values: m[1:], filterFunc: func(image v1alpha1.Image, values []string) bool {
+				return image.Spec.Builder.Kind == v1alpha1.BuilderKind && image.Spec.Builder.Name == values[0]
+			}})
+			continue
+		}
+
+		m = clusterBuilderRegex.FindStringSubmatch(flag)
+		if len(m) == 2 {
+			filters = append(filters, filter{values: m[1:], filterFunc: func(image v1alpha1.Image, values []string) bool {
+				return image.Spec.Builder.Kind == v1alpha1.ClusterBuilderKind && image.Spec.Builder.Name == values[0]
+			}})
+			continue
+		}
+
+		m = latestReasonRegex.FindStringSubmatch(flag)
+		if len(m) == 2 {
+			filters = append(filters, filter{values: strings.Split(m[1], ","), filterFunc: matchesLatestReason})
+			continue
+		}
+
+		m = statusRegex.FindStringSubmatch(flag)
+		if len(m) == 2 {
+			filters = append(filters, filter{values: strings.Split(m[1], ","), filterFunc: matchesStatus})
+			continue
+		}
+
+		return nil, fmt.Errorf(`invalid filter argument "%s"`, flag)
+	}
+
+	return filters, nil
+}
+
+func matchesAll(image v1alpha1.Image, fs []filter) bool {
+	for _, f := range fs {
+		if !f.filterFunc(image, f.values) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func matchesStatus(image v1alpha1.Image, values []string) bool {
+	contains := func(q corev1.ConditionStatus) bool {
+		for _, v := range values {
+			if strings.ToLower(string(q)) == strings.ToLower(v) {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	cond := image.Status.GetCondition(corev1alpha1.ConditionReady)
+
+	switch cond {
+	case nil:
+		return contains(corev1.ConditionUnknown)
+	default:
+		return contains(cond.Status)
+	}
+}
+
+func matchesLatestReason(image v1alpha1.Image, values []string) bool {
+	for _, v := range values {
+		if strings.ToLower(image.Status.LatestBuildReason) == strings.ToLower(v) {
```

Comment:
It seems like this this line is the first issue discussed in https://github.com/vmware-tanzu/kpack-cli/issues/141

Reply:
I think, that I agree with @sukhil-suresh that excluding images with multiple reasons seems like a regression.

Reply:
Is there any reason to not split `LatestBuildReason` by comma and check to see if any match the value?

Reply:
@sukhil-suresh Perhaps this is something that someone can take a look at with me. This is functionality that I've never seen before, therefore I don't know how it's represented in code.

To recap my understanding:
- `image.Status.LatestBuildReason` can be represented as `CONFIG+` to indicate a build by multiple reasons. I'm getting this information [from the RFC](https://github.com/pivotal/kpack/blob/master/rfcs/0004-image-filter.md).
- `image.Status.LatestBuildReason` is presented on the command-line [without any modification](https://github.com/vmware-tanzu/kpack-cli/blob/f57bd3971df8a234d3036d86d75490f3879b7d0a/pkg/commands/image/list.go#L61). 

With the above assumptions, there is not enough information to filter by `BUILDPACK`, unless something else is presented to the user (ex `CONFIG+BUILDPACK`).

Reply:
`image.Status.LatestBuildReason` should be a comma delimitated list of reasons. The logic to display the truncated reason with the `+` is [client side](https://github.com/vmware-tanzu/kpack-cli/blob/f57bd3971df8a234d3036d86d75490f3879b7d0a/pkg/commands/build/helpers.go#L38) on the build list. 

Historically we have not displayed an Image's LatestBuildReason on an image list.

Reply:
@matthewmcnew Thanks for the clarification. Paired w/ @sukhil-suresh on this. We decided to incorporate the change, but we did not add the truncation to the image list. This is because we want to show the user feedback that the filter worked (`CONFIG+` would omit the other reason a build was triggered). Ex.

```
$ kp image list --filter latest-reason=buildpack

NAME              READY     LATEST REASON              LATEST IMAGE                                                                                             NAMESPACE                                                                                             

mg-test-image1    True      CONFIG,BUILDPACK           gcr.io/cf-build-service-dev-219913/test/mg-test-image@sha1:13D5582884803A7755CEEDCB7478485498C4F1C4      mg-test
```